syntax = "proto3";

option java_multiple_files = true;
option java_package = "com.toy.grpc.lib";
option java_outer_classname = "TodoProto";

message CommonRequest {
}

message CommonReply {
  string message = 1;
}

service Todo {
  // GET
  rpc getTodo (GetTodoRequest) returns (GetTodoReply) {}
  rpc getAllTodo (CommonRequest) returns (stream GetTodoReply) {}
  rpc getCompletedTodo (CommonRequest) returns (stream GetTodoReply) {}
  rpc getUncompletedTodo (CommonRequest) returns (stream GetTodoReply) {}
  // POST
  rpc createTodo (CreateTodoRequest) returns (CreateTodoReply) {}
  // PUT
  rpc updateTodo (UpdateTodoRequest) returns (GetTodoReply) {}
  // PATCH
  rpc updateCompletion (GetTodoRequest) returns (GetTodoReply) {}
  // DELETE
  rpc deleteTodo (GetTodoRequest) returns (CommonReply) {}
  rpc deleteCompletedTodo (CommonRequest) returns (CommonReply) {}
  rpc deleteUncompletedTodo (CommonRequest) returns (CommonReply) {}
}

// POST
message CreateTodoRequest {
  string title = 1;
  string description = 2;
}

message CreateTodoReply {
  string message = 1;
  int64 id = 2;
  string title = 3;
  string description = 4;
}

// Get
message GetTodoRequest {
  int64 id = 1;
}

message GetTodoReply {
  int64 id = 1;
  string title = 2;
  string description = 3;
  bool completion = 4;
}

// PUT
message UpdateTodoRequest {
  int64 id = 1;
  string title = 2;
  string description = 3;
}